### PR TITLE
Fix #3275 in 2.12 (proposing micro-release 2.12.7.2)

### LINKED ISSUE #3275: JDK 16 Illegal reflective access for `Throwable.setCause()` with `PropertyNamingStrategy.UPPER_CAMEL_CASE`
**Describe the bug**
Mapping a json string to an instance of RuntimeException with JDK 16 (which defaults to denying illegal reflective access) while using UPPER_CAMEL_CASE property naming strategy, fails with:

```
Exception in thread "main" com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Failed to call `setAccess()` on Method 'setCause' due to `java.lang.reflect.InaccessibleObjectException`, problem: Unable to make final void java.lang.Throwable.setCause(java.lang.Throwable) accessible: module java.base does not "opens java.lang" to unnamed module
```

This manifests in parts of the AWS v1 SDK when an error is received from the AWS API endpoints.  Note example report: https://github.com/FasterXML/jackson-databind/issues/2464#issuecomment-856996757

The simplified reproduce case below is based on what the failing AWS SDK code is doing internally in `com.amazonaws.transform.JsonErrorUnmarshaller.unmarshall`.

**Version information**
2.12.3, 2.12.5, 2.13.0-rc2

**To Reproduce**
Execute this code without any `--add-opens` params and with default `--illegal-access` setting of deny.

```java
public class Test {
	public static void main(String[] args) throws Exception {
		ObjectMapper mapper = new ObjectMapper();
		mapper.setPropertyNamingStrategy(PropertyNamingStrategies.UPPER_CAMEL_CASE);

		String jsonString = "{\"message\":\"This is my runtime exception message\"}";
		JsonNode jsonContent = mapper.readTree(jsonString);

		throw mapper.treeToValue(jsonContent, RuntimeException.class);
	}
}
```

**Expected behavior**
It is expected that the code above would throw a RuntimeException with a message of "This is my runtime exception message".  This is what happens when working around the issue with java parameter `--add-opens java.base/java.lang=ALL-UNNAMED`

**Additional context**
Using java parameter `--illegal-access=debug` results in the following additional info:

```
WARNING: Illegal reflective access by com.fasterxml.jackson.databind.util.ClassUtil (file:/xxxxxx/jackson-databind-2.13.0-rc2.jar) to method java.lang.Throwable.setCause(java.lang.Throwable)
	at com.fasterxml.jackson.databind.util.ClassUtil.checkAndFixAccess(ClassUtil.java:994)
	at com.fasterxml.jackson.databind.introspect.AnnotatedMember.fixAccess(AnnotatedMember.java:139)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.fixAccess(MethodProperty.java:95)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBuilder._fixAccess(BeanDeserializerBuilder.java:522)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBuilder.build(BeanDeserializerBuilder.java:373)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.buildThrowableDeserializer(BeanDeserializerFactory.java:455)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.createBeanDeserializer(BeanDeserializerFactory.java:112)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer2(DeserializerCache.java:415)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer(DeserializerCache.java:350)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:264)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
	at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:642)
	at com.fasterxml.jackson.databind.ObjectMapper._findRootDeserializer(ObjectMapper.java:4751)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4596)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2815)
	at com.fasterxml.jackson.databind.ObjectMapper.treeToValue(ObjectMapper.java:3279)
	at Test.main(Test.java:13)
```

Note that the issue does not manifest without `mapper.setPropertyNamingStrategy(PropertyNamingStrategies.UPPER_CAMEL_CASE);`

### PR DESCRIPTION
**Pull request by millems**
The [AWS SDK for Java 1.x](https://github.com/aws/aws-sdk-java) currently supports Java 7, which is probably a bad idea, but we have customers still using this JDK version. 

Jackson, quite sanely, dropped support for Java 7 in 2.13. Until we can become equally sane, we'd really appreciate a micro-release to 2.12.7.2 to fix #3275 in the deprecated 2.12 line. Without this fix, our customers will need to manually update their Jackson version to get JDK 17+ support.

If you're generous enough to accept this pull request and release 2.12.7.2, I will sing your praises.

(Cherry-pick of https://github.com/FasterXML/jackson-databind/commit/2f60d39aa2aca52d429c34d6017b733afaffc61f)

### CODE REVIEW

**Comment by JooHyukKim** [comment]
Hmm... Is it impossible to upgrade to Jackson 2.13 on AWS SDK side?
This behavioral change is branching out from later releases, which is dangerous on maintenance perspective.

**Comment by millems** [comment]
I'd be wide open to suggestions on how to avoid this from our end. We're not able to remove `setCause` from exceptions, which is what's causing the hiccup, but if we can somehow instruct the naming strategy logic to skip that field ourselves that would be an option.

Thankfully this behavior is already present in later releases, and this is just a backport, but really anything touching earlier releases is a pain, so I totally understand looking for alternatives.

**Comment by JooHyukKim** [comment]
> but if we can somehow instruct the naming strategy logic to skip that field ourselves that would be an option.

when it comes to customization, ObjectMapper is quite flexible 😆.
Maybe below ....

```java
    mapper.setPropertyNamingStrategy(new PropertyNamingStrategy() {
        // implement your own naming strategy...
    }
```
.... approach would suit your case? Or there is bunch of other configuration methods.

Or maybe @cowtowncoder has better idea?

**Comment by cowtowncoder** [comment]
@JooHyukKim As a background, I did discuss this with Amazon OSS folks earlier and indicated this would be acceptable. I can take care of the micro-patch.

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
@millems quick question: I think this would be under AWS CCLA? Do you think anything is needed wrt addendums? I think you would be covered automatically, but wanted to double-check before merging just in case.

**Comment by millems** [comment]
Yes! Thank you. It's all kosher.

**Comment by hyandell** [comment]
+1, this would be under AWS CCLA.

File: release-notes/CREDITS-2.x [hunk 1]
```diff
@@ -1324,3 +1324,8 @@ Jelle Voost (jellevoost@github)
 JoeWoo (xJoeWoo@github)
   * Reported #3139: Deserialization of "empty" subtype with DEDUCTION failed
    (2.12.4)
+
+Jason Harper (jsharper@github)
+  * Reported #3275: JDK 16 Illegal reflective access for `Throwable.setCause()` with
+   `PropertyNamingStrategy.UPPER_CAMEL_CASE`
+  (2.12.7.2)
```
Comment (pjfanning) [inline release-notes/CREDITS-2.x:1331]: can you fix the end of file - should have an empty line ?

**Comment by pjfanning** [comment]
I am not a lawyer but the code in https://github.com/FasterXML/jackson-databind/commit/2f60d39aa2aca52d429c34d6017b733afaffc61f is already licensed to this project. If someone else creates a copy of that code and submits it as a PR to another project or another branch - they still are not the authors of the code. The original author of https://github.com/FasterXML/jackson-databind/commit/2f60d39aa2aca52d429c34d6017b733afaffc61f is still the copyright holder. So at least in my head, there is no need for a CLA for this PR.

File: release-notes/CREDITS-2.x (hunk 1 above)
Comment (millems) [reply release-notes/CREDITS-2.x]: Fixed!

**Comment by cowtowncoder** [comment]
Good point @pjfanning, I think you are right.

Fwtw I asked since in this case I knew the background for PR but it's all good either way.

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
Ok, released:

* `jackson-databind` version `2.12.7.2`
* `jackson-bom` version `2.12.7.20240502` (with consistent version set including above)

and updated:

https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.12

Jars should get to Maven Central within 10-15 minutes.

**Comment by millems** [comment]
You're awesome.

**Comment by hyandell** [comment]
100% agreed, thank you Jackson devs for your attention and work :)

**Pull request merged**